### DISCUSSION TITLE
F5-TTS Implementation option

### DISCUSSION THREAD

**Discussion by cginkatronic** [open-webui/open-webui#7376] [Ideas]
Hey, the F5-TTS works pretty good. (https://github.com/SWivid/F5-TTS)
An Idea would be to implement it as a feature in the Settings -> Audio -> TTS Settings.

**Comment by Smoffyy** [comment]
I think this would be really nice, as its another local option available.

**Comment by Beastmaster23** [comment]
I agree. There's new models trying to add emotion. Zonos uses sliders. f5-tts uses different voice clips to emulate emotions. It would be nice to let the llms decide the emotion for each response then depending on the software it will run the api.

For example the current api is 
```json
{
    model: "tts-1",
    voice: "alloy",
    input: "Today is a wonderful day to build something people love!",
    response_format: "mp3",
    speed: 1
  }
  ```
  I'm thinking we add an optional emotion key. And some settings for pitch and emotion sliders if necessary.
  
  An example of the new api for emotion based tts.
  ```json
  {
    model: "tts-1",
    voice: "alloy",
    input: "Today is a wonderful day to build something people love!",
    response_format: "mp3",
    speed: 1,
   emotion: "happy" // This is a preset. This could have any name and will have all the values like pitch, clone voice reference files, emotion sliders
   // Audio settings 
   pitch: 1.3,
   ref_voice: "\path\to\file",
   emotion_settings: {
      happy: 1
      sad: 0,
     disgust: 0.3,
     other: 0,
     anger: 0,
     surprise: .4,
     neutral: .3,
   }
  }
```

**Reply by Beastmaster23** [reply]
I was thinking about it maybe just add a pipeline so people can install different tts models for the time being especially since the tts software hasn't decided the best way for emotions to be portrayed. I didn't see if pipelines can intercept audio requests. If it can then it should be possible to make pipelines for each type tts or one that is compatible with multiple emotional tts. 
I was thinking it would take the text from the ai then predict the emotion it's trying to convey. Then convert to settings for each tts software and send the response to the user.

**Comment by Beastmaster23** [comment]
Ok I got a bandaid solution for those who want emotional tts or even use any new tts. 
Overview:
Open WebUI has an option to use open ai service so instead create a local server that handles the api call and return an audio file data. In the server you essentially wrap other api calls. If you want to do my idea above then you need to connect ollama and zonos (gradio client for your service). You need to take the text and send to ollama for sentiment analysis. For the each emotion set the correct settings to your tts then send the file as response. I can get into more detail, but I am still developing

**Comment by Beastmaster23** [comment]
Ok I did it this works for zonos. https://github.com/Beastmaster23/git-etts

**Reply by illnesse** [reply]
Neat, i might join you on this. trying to make it work with f5-tts atm

**Reply by Nitroedge** [reply]
> Ok I did it this works for zonos. https://github.com/Beastmaster23/git-etts

Do you have a demo video of it working with OpenWebUI?
Once the local server is setup as per your instructions and configured as an OpenAI endpoint in the TTS settings inside OpenWebUI, is everything else handled and it just works? 

Or is custom work needed to pass all the parameters and it doesn't just work out of the box?
Thx!

**Reply by Beastmaster23** [reply]
I can make a video but you should just try it. It's a proof of concept and a hacky way to get it to work. But yes it does everything with normal tts. It will automatically output different emotions based on your profile or the default emotions. The default is happy, sad, and neutral. I can go into more detail but it's probably better to message in my repository. This is only useful because webui still don't have support for different tts natively. Thanks for the interest.
